tools/enso4igv/src/main/java/org/enso/tools/enso4igv/EnsoSbtClassPathProvider.java

```diff
@@ -0,0 +1,155 @@
+package org.enso.tools.enso4igv;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Properties;
+import javax.swing.event.ChangeListener;
+import org.netbeans.api.java.classpath.ClassPath;
+import org.netbeans.api.java.classpath.GlobalPathRegistry;
+import org.netbeans.api.java.platform.JavaPlatform;
+import org.netbeans.spi.java.classpath.ClassPathProvider;
+import org.netbeans.spi.java.classpath.support.ClassPathSupport;
+import org.netbeans.spi.java.project.support.ProjectPlatform;
+import org.netbeans.spi.java.queries.CompilerOptionsQueryImplementation;
+import org.netbeans.spi.java.queries.SourceLevelQueryImplementation2;
+import org.netbeans.spi.project.ui.ProjectOpenedHook;
+import org.openide.filesystems.FileObject;
+import org.openide.filesystems.FileUtil;
+import org.openide.util.Exceptions;
+
+final class EnsoSbtClassPathProvider extends ProjectOpenedHook
+implements ClassPathProvider, SourceLevelQueryImplementation2, CompilerOptionsQueryImplementation {
+    private static final String BOOT = "classpath/boot";
+    private static final String SOURCE = "classpath/source";
+    private static final String COMPILE = "classpath/compile";
+    private final EnsoSbtProject project;
+    private final EnsoSources info;
+
+    EnsoSbtClassPathProvider(EnsoSbtProject prj) {
+        this.project = prj;
+        this.info = computeSbtClassPath(prj);
+    }
+
+    @Override
+    public ClassPath findClassPath(FileObject file, String type) {
+        if (FileUtil.isParentOf(project.getProjectDirectory(), file)) {
+            return switch (type) {
+                case COMPILE -> info.cp;
+                case BOOT -> info.platform.getBootstrapLibraries();
+                default -> null;
+            };
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public void projectOpened() {
+        GlobalPathRegistry.getDefault().register(SOURCE, new ClassPath[] { info.cp });
+    }
+
+    @Override
+    public void projectClosed() {
+        GlobalPathRegistry.getDefault().unregister(SOURCE, new ClassPath[] { info.cp });
+    }
+
+    private static EnsoSources computeSbtClassPath(EnsoSbtProject prj) {
+        var platform = JavaPlatform.getDefault();
+        var roots = new LinkedHashSet<>();
+        var source = "11";
+        var options = new ArrayList<String>();
+        for (FileObject ch : prj.getProjectDirectory().getChildren()) {
+            if (ch.getNameExt().startsWith(".enso-sources")) {
```

Comment:
It would be good to have those constants defined in one place but I can see how sharing that with sbt's project will be a pain.

Reply:
Indeed, these are separate compilations. Sharing a constant between them is not easy. Easier to rely on [magic strings](http://wiki.apidesign.org/wiki/MagicalStrings)...

But yes, the naming and structure of `.enso-sources` is an important API! It is a communication channel between the compiler and the _Enso4Igv_ tool. Where the file should be located (here or `target` folder or elsewhere)? What should be the format (is `java.util.Properties` really the right one)? Those are important questions that form the synergy between the pieces of the system. We have to answer them in this review, document them and stick to them in the future - changing the design would require changes at a number of places and that'd be painful.